File: shared/race/stopConfidence.ts
export type StopConfidenceLevel = "low" | "needs_review" | "high";

export interface StopConfidenceInput {
  verificationStatus: "verified" | "unverified" | "needs_review" | "rejected" | "deferred" | "pending";
  verifiedAt?: string | null;
  poiScore?: number | null;
  openingHours?: string | null;
  website?: string | null;
  phone?: string | null;
}

export interface StopConfidenceResult {
  score: number;
  level: StopConfidenceLevel;
  label: string;
}

const MS_PER_DAY = 86_400_000;
const VERIFICATION_FRESH_DAYS = 30;
const VERIFICATION_STALE_DAYS = 90;

function verificationAgeFactor(verifiedAt: string | null | undefined): number {
  if (!verifiedAt) {
    return 0;
  }
  const ageMs = Date.now() - new Date(verifiedAt).getTime();
  if (!Number.isFinite(ageMs) || ageMs < 0) {
    return 0.5;
  }
  const ageDays = ageMs / MS_PER_DAY;
  if (ageDays <= VERIFICATION_FRESH_DAYS) {
    return 1;
  }
  if (ageDays >= VERIFICATION_STALE_DAYS) {
    return 0.35;
  }
  const span = VERIFICATION_STALE_DAYS - VERIFICATION_FRESH_DAYS;
  const progress = (ageDays - VERIFICATION_FRESH_DAYS) / span;
  return 1 - progress * 0.65;
}

function poiDataScore(input: StopConfidenceInput): number {
  let score = 0;
  const poi = input.poiScore;
  if (poi != null && Number.isFinite(poi)) {
    score += Math.min(35, (poi / 100) * 35);
  }
  if (input.openingHours?.trim()) {
    score += 12;
  }
  if (input.website?.trim()) {
    score += 8;
  }
  if (input.phone?.trim()) {
    score += 5;
  }
  return score;
}

function levelFromScore(score: number, status: StopConfidenceInput["verificationStatus"]): StopConfidenceLevel {
  if (status === "needs_review" || status === "rejected") {
    return "needs_review";
  }
  if (score >= 70) {
    return "high";
  }
  if (score >= 40) {
    return "needs_review";
  }
  return "low";
}

function labelForLevel(level: StopConfidenceLevel): string {
  switch (level) {
    case "high":
      return "High confidence";
    case "needs_review":
      return "Needs review";
    default:
      return "Low confidence";
  }
}

/** Compute stop confidence from verification age, confirmations, and POI data quality. */
export function computeStopConfidence(input: StopConfidenceInput): StopConfidenceResult {
  let score = poiDataScore(input);

  if (input.verificationStatus === "verified") {
    score += 30 * verificationAgeFactor(input.verifiedAt);
  } else if (input.verificationStatus === "needs_review") {
    score += 10;
  } else if (input.verificationStatus === "pending") {
    score += 8;
  } else if (input.verificationStatus === "deferred") {
    score += 5;
  }

  score = Math.round(Math.min(100, Math.max(0, score)));
  const level = levelFromScore(score, input.verificationStatus);

  return {
    score,
    level,
    label: labelForLevel(level),
  };
}

export function stopConfidenceBadgeClass(level: StopConfidenceLevel, dark = false): string {
  if (level === "high") {
    return dark ? "bg-emerald-500/20 text-emerald-300" : "bg-emerald-50 text-emerald-700";
  }
  if (level === "needs_review") {
    return dark ? "bg-amber-500/20 text-amber-200" : "bg-amber-50 text-amber-800";
  }
  return dark ? "bg-red-500/15 text-red-300" : "bg-red-50 text-red-700";
}

File: src/stop_confidence.py
"""Stop confidence scoring (mirrors shared/race/stopConfidence.ts)."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

MS_PER_DAY = 86_400_000
VERIFICATION_FRESH_DAYS = 30
VERIFICATION_STALE_DAYS = 90


def _verification_age_factor(verified_at: str | None) -> float:
    if not verified_at:
        return 0.0
    try:
        parsed = datetime.fromisoformat(verified_at.replace("Z", "+00:00"))
        if parsed.tzinfo is None:
            parsed = parsed.replace(tzinfo=timezone.utc)
        age_ms = (datetime.now(timezone.utc) - parsed).total_seconds() * 1000
    except ValueError:
        return 0.5
    if age_ms < 0:
        return 0.5
    age_days = age_ms / MS_PER_DAY
    if age_days <= VERIFICATION_FRESH_DAYS:
        return 1.0
    if age_days >= VERIFICATION_STALE_DAYS:
        return 0.35
    span = VERIFICATION_STALE_DAYS - VERIFICATION_FRESH_DAYS
    progress = (age_days - VERIFICATION_FRESH_DAYS) / span
    return 1 - progress * 0.65


def _poi_data_score(
    *,
    poi_score: float | None,
    opening_hours: str | None,
    website: str | None,
    phone: str | None,
) -> float:
    score = 0.0
    if poi_score is not None:
        score += min(35.0, (float(poi_score) / 100.0) * 35.0)
    if opening_hours and opening_hours.strip():
        score += 12.0
    if website and website.strip():
        score += 8.0
    if phone and phone.strip():
        score += 5.0
    return score


def compute_stop_confidence(
    *,
    verification_status: str,
    verified_at: str | None = None,
    poi_score: float | None = None,
    opening_hours: str | None = None,
    website: str | None = None,
    phone: str | None = None,
) -> dict[str, Any]:
    score = _poi_data_score(
        poi_score=poi_score,
        opening_hours=opening_hours,
        website=website,
        phone=phone,
    )
    if verification_status == "verified":
        score += 30.0 * _verification_age_factor(verified_at)
    elif verification_status == "needs_review":
        score += 10.0
    elif verification_status == "pending":
        score += 8.0
    elif verification_status == "deferred":
        score += 5.0

    score = round(min(100.0, max(0.0, score)))
    if verification_status in ("needs_review", "rejected"):
        level = "needs_review"
    elif score >= 70:
        level = "high"
    elif score >= 40:
        level = "needs_review"
    else:
        level = "low"

    labels = {
        "high": "High confidence",
        "needs_review": "Needs review",
        "low": "Low confidence",
    }
    return {"score": int(score), "level": level, "label": labels[level]}


def is_high_confidence_stop(
    *,
    verification_status: str,
    verified_at: str | None = None,
    poi_score: float | None = None,
    opening_hours: str | None = None,
    website: str | None = None,
    phone: str | None = None,
) -> bool:
    result = compute_stop_confidence(
        verification_status=verification_status,
        verified_at=verified_at,
        poi_score=poi_score,
        opening_hours=opening_hours,
        website=website,
        phone=phone,
    )
    return result["level"] == "high"
